// @irys/web-upload/dist/cjs/tokens/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseWebToken = void 0;
exports.getRedstonePrice = getRedstonePrice;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const upload_core_1 = require("@irys/upload-core");
class BaseWebToken {
    constructor(config) {
        this.minConfirm = 5;
        this.isSlow = false;
        this.needsFee = true;
        this.inheritsRPC = false;
        Object.assign(this, config);
        this.config = config;
    }
    // common methods
    get address() {
        return this._address;
    }
    async ready() {
        this._address = this.wallet ? this.ownerToAddress(await this.getPublicKey()) : undefined;
    }
    async price() {
        return getRedstonePrice(this.ticker);
    }
}
exports.BaseWebToken = BaseWebToken;
exports.default = BaseWebToken;
async function getRedstonePrice(token) {
    const res = await axios_1.default.get(`https://api.redstone.finance/prices?symbol=${token}&provider=redstone&limit=1`);
    await upload_core_1.Utils.checkAndThrow(res, "Getting price data");
    return res.data[0].value;
}
