// @irys/web-upload/dist/cjs/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseWebIrys = void 0;
const tslib_1 = require("tslib");
require("@irys/upload-core/hack");
const upload_core_1 = require("@irys/upload-core");
const bundles = tslib_1.__importStar(require("./utils"));
const upload_1 = require("./upload");
class BaseWebIrys extends upload_core_1.Irys {
    // static async build({
    //   url,
    //   network,
    //   wallet,
    //   config,
    //   getTokenConfig,
    // }: {
    //   network?: Network;
    //   url?: string;
    //   wallet?: { rpcUrl?: string; name?: string; provider: object };
    //   config?: IrysConfig;
    //   getTokenConfig: (irys: BaseWebIrys) => Resolvable<WebToken>;
    // }) {
    //   // @ts-expect-error types
    //   super({ url, network, arbundles });
    //   this.debug = config?.debug ?? false;
    //   this.api = new Api({
    //     url: this.url,
    //     timeout: config?.timeout ?? 100000,
    //     headers: config?.headers,
    //   });
    // }
    constructor({ url, wallet, config, getTokenConfig, }) {
        // @ts-expect-error types
        super({ url, bundles });
        this.debug = config?.debug ?? false;
        this.api = new upload_core_1.Api({
            url: this.url,
            timeout: config?.timeout ?? 100000,
            headers: config?.headers,
        });
        this.getTokenConfig = getTokenConfig;
    }
    // todo: redo this part of the API
    async build({ wallet, config }) {
        this.tokenConfig = await this.getTokenConfig(this);
        if (this.url.host.includes("devnet.irys.xyz") && !(config?.providerUrl || wallet?.rpcUrl || this?.tokenConfig?.inheritsRPC))
            throw new Error(`Using ${this.url.host} requires a dev/testnet RPC to be configured! see https://docs.irys.xyz/build/d/networks#connecting-to-devnet`);
        this.token = this.tokenConfig.name;
        this.utils = new upload_core_1.Utils(this.api, this.token, this.tokenConfig);
        this.uploader = new upload_1.WebUploader(this);
        this.funder = new upload_core_1.Fund(this.utils);
        this.uploader = new upload_1.WebUploader(this);
        this.transactions = new upload_core_1.Transaction(this);
        this.approval = new upload_core_1.Approval(this);
        this.address = "Please run `await Irys.ready()`";
        this.uploadFolder = this.uploader.uploadFolder.bind(this.uploader);
        this.uploadFile = this.uploader.uploadFile.bind(this.uploader);
    }
}
exports.BaseWebIrys = BaseWebIrys;
exports.default = BaseWebIrys;
